src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs

```diff
@@ -131,6 +158,30 @@ public async Task Real_Client_Server_Communication()
             // We require JWT-authenticated requests to be sent over HTTPS
             var unsecuredException = await Assert.ThrowsAsync<ClientCredentialsException>(() => invoicesReadHttpClient.GetStringAsync("http://invoice-app.local/public", cts.Token));
             Assert.Equal("Due to security concerns, authenticated requests must be sent over HTTPS", unsecuredException.Message);
+
+            // Ensure the token is the same than the one we got after the background caching
+            var tokenAfterAuthenticatedRequest = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
+            Assert.NotNull(tokenAfterAuthenticatedRequest);
+            Assert.Equal(tokenAfterStartupBackgroundCaching, tokenAfterAuthenticatedRequest);
+
+            // Wait until token is expired - before that, there should be no background refresh
+            var tokenManagementService = webApp.Services.GetRequiredService<ClientCredentialsTokenManagementService>();
+            Assert.Equal(0, tokenManagementService.BackgroundRefreshedTokenCount);
+            await Task.Delay(tokenAfterAuthenticatedRequest.Expiration - DateTimeOffset.UtcNow, cts.Token);
+
+            // At this point we're not making any new authenticated HTTP request,
+            // but there should be a background refresh task that should have retrieved a new token for us and it's not the same as the one we got after the background caching
+            Assert.Equal(1, tokenManagementService.BackgroundRefreshedTokenCount);
+            var tokenAfterFirstBackgroundRefresh = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
+            Assert.NotNull(tokenAfterFirstBackgroundRefresh);
+            Assert.NotEqual(tokenAfterStartupBackgroundCaching, tokenAfterFirstBackgroundRefresh);
+
+            // If we wait a little bit longer, the token should be refreshed again
+            await Task.Delay(tokenAfterFirstBackgroundRefresh.Expiration - DateTimeOffset.UtcNow, cts.Token);
```

Comment:
I don't know how reliable this is (more generally, `Task.Delay` in a test). I wonder if you should wait for an event or a SemaphoreSlim as you did with `WaitForTokenCachingToCompleteAsync`

Reply:
I know timing isn't always reliable in tests... however here we really depend on how long tokens issued by IdentityServer are valid, and I didn't find a way to have a "hook" on their side to detect tokens are actually expired.

I tried to mitigate the issue by waiting way longer than needed (token is evicted from cache after X seconds but we wait X + eviction buffer Y which is 3 seconds + token is actually refreshed prior to this -2 seconds before)